src/algebra/big_operators/finprod.lean

```diff
@@ -0,0 +1,623 @@
+
+/-!
+### Distributivity w.r.t. addition, subtraction, and (scalar) multiplication
+-/
+
+/-- If the multiplicative supports of `f` and `g` are finite, then the product of `f i * g i` equals
+the product of `f i` multiplied by the product over `g i`. -/
+@[to_additive] lemma finprod_mul_distrib (hf : (mul_support f).finite)
+  (hg : (mul_support g).finite) :
+  ∏ᶠ i, (f i * g i) = (∏ᶠ i, f i) * ∏ᶠ i, g i :=
+begin
+  rw [finprod_eq_prod_of_mul_support_to_finset_subset _ hf (finset.subset_union_left _ _),
+    finprod_eq_prod_of_mul_support_to_finset_subset _ hg (finset.subset_union_right _ _),
+    ← finset.prod_mul_distrib],
+  refine finprod_eq_prod_of_mul_support_subset _ _,
+  simp [mul_support_mul]
+end
+
+/-- A more general version of `finprod_mem_mul_distrib` that requires `s ∩ mul_support f` and
+`s ∩ mul_support g` instead of `s` to be finite. -/
+@[to_additive] lemma finprod_mem_mul_distrib' (hf : (s ∩ mul_support f).finite)
+  (hg : (s ∩ mul_support g).finite) :
+  ∏ᶠ i ∈ s, (f i * g i) = (∏ᶠ i ∈ s, f i) * ∏ᶠ i ∈ s, g i :=
+begin
+  rw [← mul_support_mul_indicator] at hf hg,
+  simp only [finprod_mem_def, mul_indicator_mul, finprod_mul_distrib hf hg]
+end
+
+/-- The product of constant one over any set equals one. -/
+@[to_additive] lemma finprod_mem_one (s : set α) : ∏ᶠ i ∈ s, (1 : M) = 1 := by simp
+
+/-- If a function `f` equals one on a set `s`, then the product of `f i` over `i ∈ s` equals one. -/
+@[to_additive] lemma finprod_mem_of_eq_on_one (hf : eq_on f 1 s) : ∏ᶠ i ∈ s, f i = 1 :=
+by { rw ← finprod_mem_one s, exact finprod_mem_congr rfl hf }
+
+/-- If the product of `f i` over `i ∈ s` is not equal to one, then there is some `x ∈ s`
+such that `f x ≠ 1`. -/
+@[to_additive] lemma exists_ne_one_of_finprod_mem_ne_one (h : ∏ᶠ i ∈ s, f i ≠ 1) :
+  ∃ x ∈ s, f x ≠ 1 :=
+begin
+  by_contra h', push_neg at h',
+  exact h (finprod_mem_of_eq_on_one h')
+end
+
+/-- Given a finite set `s`, the product of `f i * g i` over `i ∈ s` equals the product of `f i`
+over `i ∈ s` times the product of `g i` over `i ∈ s`. -/
+@[to_additive] lemma finprod_mem_mul_distrib (hs : s.finite) :
+  ∏ᶠ i ∈ s, (f i * g i) = (∏ᶠ i ∈ s, f i) * ∏ᶠ i ∈ s, g i :=
+finprod_mem_mul_distrib' (hs.inter_of_left _) (hs.inter_of_left _)
+
+@[to_additive] lemma finprod_hom {f : α → M} (g : M →* N) (hf : (mul_support f).finite) :
+  ∏ᶠ i, g (f i) = g (∏ᶠ i, f i) :=
+begin
+  rw [finprod_eq_prod _ hf, g.map_prod],
+  refine finprod_eq_prod_of_mul_support_subset _ _,
+  simp [mul_support_comp_subset g.map_one]
+end
+
+/-- A more general version of `finprod_mem_hom` that requires `s ∩ mul_support f` and instead of
+  `s` to be finite. -/
+@[to_additive] lemma finprod_mem_hom' {f : α → M} (g : M →* N) (h₀ : (s ∩ mul_support f).finite) :
+  ∏ᶠ i ∈ s, (g (f i)) = g (∏ᶠ j ∈ s, f j) :=
+begin
+  rw [finprod_mem_def, mul_indicator_comp_of_one g.map_one, finprod_hom, finprod_mem_def],
+  rwa mul_support_mul_indicator
+end
```

Comment:
For `finset.prod` we call this `monoid_hom.map_prod` and flip the direction of the equality. I think it would make sense to do the same thing here, and call it `monoid_hom.map_finprod_mem'`?